internal/repository/ssh.go

```diff
@@ -1,257 +1,543 @@
+
+			agentSigners, err := agentClient.Signers()
+			if err != nil {
+				log.Debug("request signers from agent failed: " + err.Error())
+				return explictSigners, nil
+			}
+
+			return append(agentSigners, explictSigners...), nil
+		})
+
+		authMethods = append(authMethods, authPublickey)
+	}
+
+	// setup the ssh client and sftp client
+	var sshClient *ssh.Client
+	sshConfig := ssh.ClientConfig{
+		User:            user,
+		Auth:            authMethods,
+		HostKeyCallback: hostkeyCallback,
 	}
+	if proxyCommand != "" {
+		proxyCommand = strings.ReplaceAll(proxyCommand, "%h", hostname)
+		proxyCommand = strings.ReplaceAll(proxyCommand, "%p", strconv.Itoa(port))
 
-	taskRemote := func(ctx context.Context) error {
-		var stderr = bytes.Buffer{}
-		cmd.Stderr = &stderr
-		err = cmd.Run()
+		proxyCommandConn, err := newProxyCommandConn(proxyCommand)
 		if err != nil {
-			return errors.New(stderr.String())
+			return nil, err
 		}
 
-		return nil
+		c, chans, reqs, err := ssh.NewClientConn(proxyCommandConn, hostname, &sshConfig)
+		if err != nil {
+			return nil, err
+		}
+
+		sshClient = ssh.NewClient(c, chans, reqs)
+		log.Debugf("connect to %s@%s successfully\n", user, hostname)
+	} else if proxyJump != "" {
+		proxyClient, err := newSSHClient(proxyJump, true)
+		if err != nil {
+			return nil, err
+		}
+
+		proxyJumpConn, err := proxyClient.Dial("tcp", fmt.Sprintf("%s:%d", hostname, port))
+		if err != nil {
+			return nil, err
+		}
+
+		ncc, chans, reqs, err := ssh.NewClientConn(proxyJumpConn, fmt.Sprintf("%s:%d", hostname, port), &sshConfig)
+		if err != nil {
+			return nil, err
+		}
+
+		sshClient = ssh.NewClient(ncc, chans, reqs)
+		log.Debugf("connect to %s@%s successfully\n", user, hostname)
+	} else {
+		sshClient, err = ssh.Dial("tcp", fmt.Sprintf("%s:%d", hostname, port), &sshConfig)
+		if err != nil {
+			return nil, err
+		}
+		log.Debugf("connect to %s@%s at port %d successfully\n", user, hostname, port)
+	}
+
+	return sshClient, nil
+}
+
+func sshLoadIdentifyFile(identityFile string) (ssh.Signer, error) {
+	key, err := ioutil.ReadFile(normalizeKeyPath(identityFile))
+	if err != nil {
+		return nil, err
 	}
 
-	err = executor.ExecuteAll(2, taskLocal, taskRemote)
+	var signer ssh.Signer
+	passphrase := os.Getenv("SSH_KEY_PASSPHRASE")
+	if passphrase == "" {
+		signer, err = ssh.ParsePrivateKey(key)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		signer, err = ssh.ParsePrivateKeyWithPassphrase(key, []byte(passphrase))
+		if err != nil {
+			return nil, err
+		}
+	}
+	return signer, nil
+}
+
+func sshKnownhostCallback(hostname string, remote net.Addr, key ssh.PublicKey) error {
+	currentUser, err := osuser.Current()
 	if err != nil {
 		return err
 	}
 
+	knownHostFile := filepath.Join(currentUser.HomeDir, ".ssh", "known_hosts")
+	log.Debugln("check known hosts by file " + knownHostFile)
+
+	knownhostCallback, err := knownhosts.New(knownHostFile)
+	if err != nil {
+		log.Debug("cannot load knownhost file: " + err.Error())
```

Comment:
should it return `err` here?

Reply:
No. It is just warning. known_hosts file could be absent.